### PR TITLE
Ports TG IV Drips

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Updates our TG drips to TG's

- Different sprites
- Icon changes depending on whether the mode is injecting or taking blood
- Animations when it's actually injecting/taking blood
- Transfer rate increased (it's faster to syringe blood out of a beaker and into someone with current IV's)
- Warning ping when someone is low on blood
- Support for species with exotic blood
- Alt-Click to toggle between modes

Animation

img

Code
- spanssss
- general clean up
- IV drips are no longer in machine processing if they're not being used; they are added when they're attached and removed when they're no longer needed
- Should GC beter now

:cl: Fox McCloud
add: IV Drip sprites updated; has visual feedback for if it's injecting or taking blood
tweak: IV Drip injection rate dramatically increased
add: Warning ping for when someone is low on blood
add: Support for species with exotic blood type
add: Alt-Click to toggle between modes
/:cl:

### CODE REVIEW

**Comment by Krausus** [comment]
> Alt-Click to toggle between modes

I made a comment about alt-clicking on the other PR that uses it:

> Oh, geez, don't add new clicks that work like this. It means trying to alt-click something in the world to manipulate it will hijack your current stat tab, and trying to alt-click something in the "turf contents" tab will close it.

> Alt-click is fine for stuff in the interface, such as the PDA ID removal, but stuff in-world should use a click combo that's not already used. Ctrl-click would probably work fine, or, since you're replacing verbs you'd normally right-click for, you could even do something like alt-right-click.

> But please, don't put alt-click on double duty.

File: code/game/machinery/iv_drip.dm [hunk 1]
```diff
@@ -1,101 +1,128 @@
 /obj/machinery/iv_drip
 	name = "\improper IV drip"
 	icon = 'icons/obj/iv_drip.dmi'
+	icon_state = "iv_drip"
 	anchored = 0
 	density = 1
+	var/mob/living/carbon/human/attached = null
+	var/mode = 1 // 1 is injecting, 0 is taking blood.
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: Rename this variable to something more obvious like `injecting`, *or* use defines.

File: code/game/machinery/iv_drip.dm [hunk 2]
```diff
@@ -1,101 +1,128 @@
 /obj/machinery/iv_drip
 	name = "\improper IV drip"
 	icon = 'icons/obj/iv_drip.dmi'
+	icon_state = "iv_drip"
 	anchored = 0
 	density = 1
+	var/mob/living/carbon/human/attached = null
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: Explicitly setting `null` seems redundant.

File: code/game/machinery/iv_drip.dm [hunk 3]
```diff
@@ -1,101 +1,128 @@
 /obj/machinery/iv_drip
 	name = "\improper IV drip"
 	icon = 'icons/obj/iv_drip.dmi'
+	icon_state = "iv_drip"
 	anchored = 0
 	density = 1
+	var/mob/living/carbon/human/attached = null
+	var/mode = 1 // 1 is injecting, 0 is taking blood.
+	var/obj/item/weapon/reagent_containers/beaker = null
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: Again.

File: code/game/machinery/iv_drip.dm [hunk 4]
```diff
@@ -1,101 +1,128 @@
...
+		if(beaker)
+			to_chat(user, "<span class='warning'>There is already a reagent container loaded!</span>")
 			return
-
-		if(user.drop_item())
-			W.forceMove(src)
-			src.beaker = W
-			to_chat(user, "You attach \the [W] to \the [src].")
-			src.update_icon()
+		if(!user.drop_item())
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: Can't you make `drop_item` drop directly into the IV with some arguments?

File: code/game/machinery/iv_drip.dm [hunk 5]
```diff
@@ -1,101 +1,128 @@
...
 			return
-		else
-			to_chat(user, "\The [W] is stuck to you!")
+
+		W.forceMove(src)
+		beaker = W
+		to_chat(user, "<span class='notice'>You attach [W] to [src].</span>")
+		update_icon()
+		return
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: This return is redundant.

File: code/game/machinery/iv_drip.dm [hunk 6]
```diff
@@ -1,101 +1,128 @@
...
+	if(!attached)
+		return PROCESS_KILL
 
-		if(!(get_dist(src, src.attached) <= 1 && isturf(src.attached.loc)))
-			visible_message("The needle is ripped out of [src.attached], doesn't that hurt?")
-			src.attached:apply_damage(3, BRUTE, pick("r_arm", "l_arm"))
-			src.attached = null
-			src.update_icon()
-			return
+	if(!(get_dist(src, attached) <= 1 && isturf(attached.loc)))
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: `get_dist(src, attached) > 1`

File: code/game/machinery/iv_drip.dm [hunk 7]
```diff
@@ -104,71 +131,111 @@
...
+		return
+	toggle_mode()
 
+/obj/machinery/iv_drip/verb/eject_beaker(mob/user)
+	set category = "Object"
+	set name = "Remove IV Container"
+	set src in view(1)
+
+	if(!iscarbon(usr))
+		usr << "<span class='warning'>You can't do that!</span>"
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: Gonna need `to_chat`

File: code/game/machinery/iv_drip.dm (hunk 4 above)
Comment (Fox-McCloud) [inline code/game/machinery/iv_drip.dm]: Uh, no. `drop_item` doesn't take arguments

File: code/game/machinery/iv_drip.dm [hunk 8]
```diff
@@ -104,71 +131,111 @@
...
+		return
+
+	if(usr.incapacitated())
```
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: No adjacency check?

File: code/game/machinery/iv_drip.dm (hunk 4 above)
Comment (PJB3005) [inline code/game/machinery/iv_drip.dm]: Guess that's a /vg/ thing then, rip.

**Comment by Lady-Luck** [comment]
Can you make it possible for medical borgs to use these?

**Comment by Davdichan** [comment]
I know /tg/ specifically made their IV drips incompatible with monkeys and simple mods. Will medical still be able to set up blood farms with Monkeys/Farwas after this change?

File: code/game/machinery/iv_drip.dm (hunk 8 above)
Comment (Fox-McCloud) [inline code/game/machinery/iv_drip.dm]: Verb only applies to those who in the tiles adjacent to it anyway, so....

**Comment by Fox-McCloud** [comment]
@Davdichan 

TG IV drips work with monkies---that's unchanged here.

**Pull request merged**